src/network/networkd-sysctl.c

```diff
@@ -5,15 +5,190 @@
 #include <linux/if_arp.h>
 
 #include "af-list.h"
+#include "fd-util.h"
 #include "missing_network.h"
 #include "networkd-link.h"
 #include "networkd-manager.h"
 #include "networkd-network.h"
 #include "networkd-sysctl.h"
+#include "path-util.h"
 #include "socket-util.h"
 #include "string-table.h"
 #include "sysctl-util.h"
 
+#if HAVE_VMLINUX_H
+
+#include "bpf/sysctl_monitor/sysctl-monitor-skel.h"
+#include "bpf/sysctl_monitor/sysctl-write-event.h"
+
+#define CGROUP_MOUNT_DFLT "/sys/fs/cgroup"
+
+static int open_rootcg(void) {
+        int fd = open(CGROUP_MOUNT_DFLT, O_PATH | O_DIRECTORY | O_CLOEXEC);
+
+        if (fd < 0)
+                log_error_errno(errno, "Failed to open cgroup mount point %s: %m", CGROUP_MOUNT_DFLT);
+
+        return fd;
+}
+
+static int sysct_write_event(void *ctx, void *data, size_t data_sz) {
+        struct sysctl_write_event *we = ASSERT_PTR(data);
+        Hashmap **sysctl_shadow = ASSERT_PTR(ctx);
+        _cleanup_free_ char *path = NULL;
+        char *value;
+
+        /* Ignore our own writes */
+        if (we->pid == getpid())
+                return 0;
+
+        path = path_join("/proc/sys", we->name);
+        if (!path)
+                return log_oom();
+
+        /* If we never managed this handle, ignore it. */
+        value = hashmap_get(*sysctl_shadow, path);
+        if (!value)
+                return 0;
+
+        if (!streq(value, we->newvalue))
+                log_warning("'%s' changed sysctl '%s' from '%s' to '%s', we want '%s'", we->comm, we->name, we->current, we->newvalue, value);
+
+        return 0;
+}
+
+static int on_ringbuf_io(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
+        struct ring_buffer *rb = ASSERT_PTR(userdata);
+        int r;
+
+        r = ring_buffer__poll(rb, 1000);
+        if (r < 0 && errno != EINTR)
+                log_error_errno(errno, "Error polling ring buffer: %m");
+
+        return 0;
+}
+
+void sysctl_add_monitor(Manager *manager) {
+        _cleanup_close_ int cgroup_fd = -EBADF;
+        int prog_fd;
+        int r;
+
+        assert(manager);
+
+        r = dlopen_bpf();
+        if (r < 0) {
+                log_warning_errno(r, "Failed to load BPF library: %m");
```

Comment:
this really should be handled gracefully, bpf is not universally available. LOG_WARNING is too high a log level.

maybe do instead:

```
log_info("sysctl monitor disabled, as BPF support is not available.");
``